### PR TITLE
add RDS deploy

### PR DESCRIPTION
**Pull request by viacheslav-karnaukh**
**Products API root url**: https://cc3mwqo2y1.execute-api.eu-west-1.amazonaws.com/dev/products, use [swagger](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/viacheslav-karnaukh/node-aws-services/task-4/product-service/swagger-dev.json) for details
**Frontend application**: https://d57iuemqb7f9r.cloudfront.net/

- [x] 1 - Task 4.1 is implemented
- [x] 3 - TASK 4.2 is implemented lambda links are provided and returns data
- [x] 4 - TASK 4.3 is implemented lambda links are provided and products is stored in DB (call TASK 4.2 to see the product)
- [x] 5 - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. Link to a working Front-End application is provided for cross-check reviewer.
Additional (optional) tasks (but nice to have):

- [x] +1 - POST/products lambda functions returns error 400 status code if product data is invalid
- [x] +1 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] +1 - All lambdas do console.log for each incoming requests and their arguments
- [x] +1 - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa) (https://devcenter.kinvey.com/nodejs/tutorials/bl-transactional-support)

### CODE REVIEW

File: product-service/functions/addProduct.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  try {
+    const invalidFields = findInvalidFields(productData, productSchema);
+
+    if(invalidFields.length) {
+      console.log('invalid fields -->', findInvalidFields);
+      throw new ProductsModelError('Product data is invalid', 400);
+    }
+
+    const createdProduct = await insertProduct(productData);
```
Comment (Emerkof) [inline product-service/functions/addProduct.ts:41]: @viacheslav-karnaukh Can't create a product from your client app due to CORS error:

image


Haven't your forgotten to add CORS headers for this route?

File: product-service/model/productsModel.ts [hunk 2]
```diff
@@ -1,5 +1,25 @@
...
+  user: PGUSER,
+  host: PGHOST,
+  database: PGDATABASE,
+  password: PGPASSWORD,
+  port: Number(PGPORT),
+  ssl: {
+    rejectUnauthorized: false,
+  },
+  connectionTimeoutMillis: 5000,
+};
```
Comment (Emerkof) [inline product-service/model/productsModel.ts:22]: @viacheslav-karnaukh I would strongly suggest to encapsulate db connection functionality in 1 module and export this module across your app. This way, if you need to change database user/host/password, you'll only need to change it in 1 file.

It's a wish for the future, no need to implement it right now, it won't affect your score.

**Review by Emerkof — CHANGES_REQUESTED** [review]

File: product-service/model/create_samples.sql [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  id uuid primary key default uuid_generate_v4(),
+  title text not null,
+  description text,
+  price integer
+);
+
+drop table if exists stocks;
+
+create table stocks (
+  id serial primary key,
```
Comment (agarmashevska) [inline product-service/model/create_samples.sql:15]: should not be additional "id" except product_id according to requirements

Comment (Emerkof) [reply product-service/model/create_samples.sql:15]: @agarmashevska I don't think that Product or Stock tables should strictly adhere to the schemas, described in the task. As the course creators said: 'Your model/site can be whatever you want'. So it's nothing bad that he added 1 additional field to the model.

File: product-service/functions/addProduct.ts (hunk 1 above)
Comment (viacheslav-karnaukh) [reply product-service/functions/addProduct.ts:41]: thanks. seems like I broke it after some manual research in aws console. anyway, fixed it and now it works

File: product-service/model/create_samples.sql (hunk 3 above)
Comment (viacheslav-karnaukh) [reply product-service/model/create_samples.sql:15]: @agarmashevska, @Emerkof
I would say more. It is good practice and id should be present for every table in real life. But for this task we use such a simplification and it is ok not to add id for now. https://www.youtube.com/watch?v=7zd3K6zqmmE&feature=youtu.be&t=3111 - here they explain it.

**Review by petr150894 — APPROVED** [review]

**Pull request merged**